### ISSUE TITLE
Thanks for your work, but I don't see the code for the loss of the contrast alignment part, is this part not open source please?

### ISSUE THREAD

**Issue by HUAFOR** [binbinjiang/CVT-SLR#2]

**Comment by binbinjiang** [comment]
Thanks for your attention, the contrastive loss is easy to implement, which is similar to the CLIP loss. Referring to the description in our paper, I am providing the code for this loss function as follows:
`
def contra_loss(self, visual_repr, context_repr, logit_scale):
	# visual_repr => (Tensor) T,B,C
	# context_repr => (Tensor) T,B,C
	# logit_scale => (float) hyper-parameter
	
	visual_repr = visual_repr.transpose(0,1)
	context_repr = context_repr.transpose(0,1)

	bt = visual_repr.shape[0]
	assert bt == context_repr.shape[0]

	visual_repr = visual_repr.reshape(bt, -1)
	context_repr = context_repr.reshape(bt, -1)

	# normalized features
	visual_repr = visual_repr / visual_repr.norm(dim=-1, keepdim=True)
	context_repr = context_repr / context_repr.norm(dim=-1, keepdim=True)

	# cosine similarity as logits
	logit_scale = logit_scale.exp()
	logits_per_struc = logit_scale * visual_repr @ context_repr.T
	logits_per_seq = logits_per_struc.T

	labels = torch.arange(bt, device=visual_repr.device, dtype=torch.long)

	loss_struc = F.cross_entropy(logits_per_struc, labels)
	loss_seq = F.cross_entropy(logits_per_seq, labels)
	contra_loss = (loss_struc + loss_seq) / 2

	# return logits_per_struc, logits_per_seq
	return contra_loss
`

Hope it works for you! 
Best regards!

**Comment by HUAFOR** [comment]
Awesome! Thank you for your reply!
